blocks/certifier.go

```diff
@@ -0,0 +1,439 @@
+	if c.isShuttingDown() {
+		return pubsub.ValidationIgnore
+	}
+
+	err := c.handleRawCertifyMsg(ctx, msg)
+	switch {
+	case err == nil:
+		return pubsub.ValidationAccept
+	case errors.Is(err, errMalformedData) == true:
+		return pubsub.ValidationReject
+	case errors.Is(err, errKnownCertMsg) == true || errors.Is(err, errMsgTooOld):
+		return pubsub.ValidationIgnore
+	default:
+		c.logger.WithContext(ctx).With().Error("failed to process certify message gossip", log.Err(err))
+		return pubsub.ValidationIgnore
+	}
+}
+
+// NumCached returns the number of layers being cached in memory.
+func (c *Certifier) NumCached() int {
+	c.mu.Lock()
+	defer c.mu.Unlock()
+	return len(c.certMsgs)
+}
+
+func (c *Certifier) getCertState(logger log.Log, lid types.LayerID, bid types.BlockID) (certState, error) {
+	c.mu.Lock()
+	defer c.mu.Unlock()
+	return c.getCertStateLocked(logger, lid, bid)
+}
+
+func (c *Certifier) getCertStateLocked(logger log.Log, lid types.LayerID, bid types.BlockID) (certState, error) {
+	if ci, ok := c.certMsgs[lid]; !ok {
+		logger.Error("layer not registered for cert")
+		return unknown, errInternal
+	} else if ci.bid != bid {
+		logger.Error("block not registered for cert")
+		return unknown, errInternal
+	}
+
+	if time.Now().After(c.certMsgs[lid].deadline) {
+		logger.With().Debug("cert msg received after deadline", log.Time("deadline", c.certMsgs[lid].deadline))
+		return expired, nil
+	}
+	if c.certMsgs[lid].done {
+		return certified, nil
+	}
+	return pending, nil
+}
+
+// HandleSyncedCertificate handles Certificate from sync.
+func (c *Certifier) HandleSyncedCertificate(ctx context.Context, lid types.LayerID, cert *types.Certificate) error {
+	logger := c.logger.WithContext(ctx).WithFields(cert.BlockID)
+	for _, msg := range cert.Signatures {
+		if err := c.validate(ctx, logger, msg); err != nil {
+			return err
+		}
+	}
+	if err := c.save(logger, lid, cert); err != nil {
+		return err
+	}
+	return nil
+}
+
+func (c *Certifier) handleRawCertifyMsg(ctx context.Context, data []byte) error {
+	logger := c.logger.WithContext(ctx)
+	var msg types.CertifyMessage
+	if err := codec.Decode(data, &msg); err != nil {
+		logger.With().Error("malformed cert msg", log.Err(err))
+		return errMalformedData
+	}
+	if err := c.validate(ctx, logger, msg); err != nil {
+		return err
+	}
+	if err := c.saveMessage(logger, msg); err != nil {
+		return errInternal
+	}
+	return nil
+}
+
+func (c *Certifier) validate(ctx context.Context, logger log.Log, msg types.CertifyMessage) error {
+	lid := msg.LayerID
+	bid := msg.BlockID
+	if state, err := c.getCertState(logger, lid, bid); err != nil {
+		return err
+	} else if state == expired {
+		return errMsgTooOld
+	} else if state == certified {
+		// should still gossip this msg to peers even when this node has created a certificate
+		return nil
+	}
+
+	logger.WithFields(bid)
+	if bid != types.EmptyBlockID {
+		if err := c.verifyLayerID(logger, lid, bid); err != nil {
```

Comment:
A call to database for every message in a certificate?
Seems a little excessive. That's 400 successive database calls when validating a certificate. 

Especially since the certificate needs only be internally consistent to be trusted. We can take the certificate as truth. No need to check against our local understanding of the layer.

Reply:
it's not about checking local understanding of layer. the point was to verify this block is for real, that we have this block.

the concept of this network is that we only accept and gossip about msg we validate.
and a msg is only valid if everything referenced in the msg were validated.

in any case, i removed this check as i later changed the code to check it earlier at getCertState. there it was just an in-memory check.